### PR TITLE
Prevent APOs and agreements from being republished

### LINKED ISSUE #4050: Bulk action "republish" should log an error when an attempt is made to "republish" an APO or agreement
**Describe the bug**
Only collections and items can be published/republished to a Purl. APOs and agreements are for internal use only. On the show page, there's no way to publish an APO or agreement. But in the bulk actions, APOs and agreements can be listed inadvertently in a druid list when running the "republish" bulk action.

When an APO/agreement is republished via bulk action, sidekiq retries endlessly. There are 4 APOs currently in the sidekiq retries queue: https://dor-services-app-prod-a.stanford.edu/queues/retries

I think this is the HB error: https://app.honeybadger.io/projects/50568/faults/97302684

**User Impact**
It's not very noticeable to a non-admin user. The bulk action reports success on the Argo bulk actions page while the publish retries keep happening in the background.

**To Reproduce**
1. Go to the Argo Bulk actions page in stage
2. Create a new bulk action for "republish"
3. Enter an APO druid such as qg426bg6497
4. See the error in the sidekiq retries

**Expected behavior**
Since APOs and agreements can't be published, when an attempt is made to publish them via bulk action, an error should be reported to the bulk action log so the user can see that there was a problem. No retries should be attempted.

**Additional context**
Some APOs and agreements apparently have Purls, even though they're not supposed to. There was probably a bug a long time ago that allowed them to be published. I inadvertently tried to republish these without realizing they should never have been published because I'm trying to put cocina JSON out there for anything that has a purl with XML.

### PR DESCRIPTION
**Pull request by mjgiarlo**
# Why was this change made? 🤔

Fixes #4050

# How was this change tested? 🤨

- [x] CI
- [x] QA/stage

# Does your change introduce accessibility violations? 🩺

Nah.

### CODE REVIEW

File: app/jobs/republish_job.rb [hunk 1]
```diff
@@ -19,7 +19,10 @@ def perform(bulk_action_id, params)
   private
 
   def republish(current_druid, log_buffer)
-    if WorkflowService.published?(druid: current_druid)
+    if !publishable?(Repository.find(current_druid))
```
Comment (mjgiarlo) [inline app/jobs/republish_job.rb:22]: First, make sure the object is a type that can be published.

File: app/jobs/republish_job.rb [hunk 2]
```diff
@@ -31,4 +34,14 @@ def republish(current_druid, log_buffer)
...
   end
+
+  def publishable?(maybe_cocina_object)
+    return false if maybe_cocina_object.is_a?(NilModel) ||
+      maybe_cocina_object.admin_policy? ||
+      maybe_cocina_object.type == Cocina::Models::ObjectType.agreement
+
+    true
+  rescue
+    false
```
Comment (mjgiarlo) [inline app/jobs/republish_job.rb:45]: Failsafe.

File: app/jobs/republish_job.rb [hunk 3]
```diff
@@ -31,4 +34,14 @@ def republish(current_druid, log_buffer)
     log_buffer.puts("#{Time.current} #{self.class}: Unexpected error for #{current_druid} (bulk_action.id=#{bulk_action.id}): #{e}")
     bulk_action.increment(:druid_count_fail).save
   end
+
+  def publishable?(maybe_cocina_object)
```
Comment (mjgiarlo) [inline app/jobs/republish_job.rb:38]: `Repository#find` either returns a Cocina model or an instance of `NilModel`, hence the name `maybe_cocina_object` and the checks below.

File: app/jobs/republish_job.rb [hunk 4]
```diff
@@ -31,4 +34,14 @@ def republish(current_druid, log_buffer)
...
+    return false if maybe_cocina_object.is_a?(NilModel) ||
+      maybe_cocina_object.admin_policy? ||
+      maybe_cocina_object.type == Cocina::Models::ObjectType.agreement
+
+    true
```
Comment (mjgiarlo) [inline app/jobs/republish_job.rb:43]: Flag *specific* known cases and default to `true`.

File: spec/jobs/republish_job_spec.rb [hunk 5]
```diff
@@ -3,18 +3,19 @@
...
 RSpec.describe RepublishJob do
-  let(:druids) { ["druid:123", "druid:456"] }
+  let(:druid) { "druid:123" }
+  let(:druids) { Array(druid) }
   let(:groups) { [] }
   let(:user) { instance_double(User, to_s: "jcoyne85") }
   let(:bulk_action) { create(:bulk_action) }
-
-  let(:client1) { instance_double(Dor::Services::Client::Object, publish: true) }
-  let(:client2) { instance_double(Dor::Services::Client::Object, publish: true) }
```
Comment (mjgiarlo) [inline spec/jobs/republish_job_spec.rb:12]: I didn't notice any value added to this spec by testing multiple druids.

File: spec/jobs/republish_job_spec.rb [hunk 6]
```diff
@@ -3,18 +3,19 @@
...
+  let(:object) { FactoryBot.create_for_repository(:persisted_item) }
+  let(:object_client) { instance_double(Dor::Services::Client::Object, publish: true, find: object) }
+  let(:workflow_client) { instance_double(Dor::Workflow::Client, lifecycle: lifecycle) }
```
Comment (mjgiarlo) [inline spec/jobs/republish_job_spec.rb:13]: Be more specific with variable names.

File: spec/jobs/republish_job_spec.rb [hunk 7]
```diff
@@ -3,18 +3,19 @@
...
+  let(:lifecycle) { Time.zone.now }
```
Comment (mjgiarlo) [inline spec/jobs/republish_job_spec.rb:14]: Parameterize the smallest part that varies.

**Review by jmartin-sul — APPROVED** [review]

**Pull request merged**